_posts/2021-07-13-JEV.md

```diff
@@ -0,0 +1,822 @@
+          // xyzxyzxy
+          final FloatVector firstVec = FloatVector.fromArray(SPECIES, src, i + 0);
+          firstMinVec = firstVec.min(firstMinVec);
+          firstMaxVec = firstVec.max(firstMaxVec);
+
+          // zxyzxyzx
+          final FloatVector secondVec = FloatVector.fromArray(SPECIES, src, i + 8);
+          secondMinVec = secondVec.min(secondMinVec);
+          secondMaxVec = secondVec.max(secondMaxVec);
+
+          // yzxyzxyz
+          final FloatVector thirdVec = FloatVector.fromArray(SPECIES, src, i + 16);
+          thirdMinVec = thirdVec.min(thirdMinVec);
+          thirdMaxVec = thirdVec.max(thirdMaxVec);
+        }
+        
+        // we now have 8 values in total for each of x, y, z
+        // so min/max-ing them takes a little more effort 
+      
+        final float[] allMins = new float[24];
+        firstMinVec.intoArray(allMins, 0);
+        secondMinVec.intoArray(allMins, 8);
+        thirdMinVec.intoArray(allMins, 16);
+
+        final float[] allMaxs = new float[24];
+        firstMaxVec.intoArray(allMaxs, 0);
+        secondMaxVec.intoArray(allMaxs, 8);
+        thirdMaxVec.intoArray(allMaxs, 16);
+
+        for (int i = 0; i < 8; i++) {
+            // x
+            min[0] = Math.min(min[0], allMins[(i * 3) + 0]);
+            max[0] = Math.max(max[0], allMaxs[(i * 3) + 0]);
+
+            // y
+            min[1] = Math.min(min[1], allMins[(i * 3) + 1]);
+            max[1] = Math.max(max[1], allMaxs[(i * 3) + 1]);
+
+            // z
+            min[2] = Math.min(min[2], allMins[(i * 3) + 2]);
+            max[2] = Math.max(max[2], allMaxs[(i * 3) + 2]);
+        }
+    }
+}
+```
+
+So, let's do a little benchmarking, and see how we did:
+
+|vanilla      |simd         |simd-768bit  |
+|---          |---          |---          |
+|34000ns      |18500ns      |6100ns       |
+
+Holy guacamole, that's almost 3 times as fast as the simd version and almost **6 times** faster than the vanilla one!!!!
+1One
+
+How is this possible you ask? This is only conjecture, but most CPU's have 8 SSE (Streaming SIMD Extension) registers,
+meaning our 3 loads/mins/maxs are most likely done concurrently to different registers, instead of sequentially to the
+same one. Could we go up to 6 registers then...you might ask?
+
+We could, and we did, but we didn't gain any performance. Meaning that this would be an interesting subject-matter for a
+follow-up rant :) 
+
+### Epilogue
+
+The Vector API is a very exciting addition to the Java realm. At first glance it looks complicated and difficult to use,
+but it is easy to get used to, and pleasantly documented.
+
+A lot of libraries that might depend on JNI or other more native-centric techniques for performance would benefit
+drastically from the readability and maintainability that writing code with the Vector API will provide.
+
+The bigger problem lies with this new hammer to which many a programmer will seek to convert their problem into a
```

Comment:
Yea, perhaps it could be expanded on a little bit more, when SIMD is *not* the tool to use (disadvantages).

I see repetition, let's fire up our SIMD-hammer 3000!